arc_cache.go

```diff
@@ -145,20 +192,34 @@ func (b *arccache) Put(bl blocks.Block) error {
 }
 
 func (b *arccache) PutMany(bs []blocks.Block) error {
+	mxs := [256]*sync.Mutex{}
 	var good []blocks.Block
 	for _, block := range bs {
 		// call put on block if result is inconclusive or we are sure that
 		// the block isn't in storage
 		if has, _, ok := b.queryCache(block.Cid()); !ok || (ok && !has) {
 			good = append(good, block)
+			mxs[block.Cid().KeyString()[len(block.Cid().KeyString())-1]] = &sync.Mutex{}
 		}
 	}
+
+	for _, mx := range mxs {
+		if mx != nil {
+			mx.Lock()
+		}
+	}
+
 	err := b.blockstore.PutMany(good)
 	if err != nil {
 		return err
 	}
 	for _, block := range good {
 		b.cacheSize(block.Cid(), len(block.RawData()))
+		if mx := mxs[block.Cid().KeyString()[len(block.Cid().KeyString())-1]]; mx != nil {
+			mx.Unlock()
```

Comment:
We need to hold the lock till we're completely done.

Reply:
realizing now that these unlock calls will need to happen in a deferred function in case `err := b.blockstore.PutMany(good)` returns an error.